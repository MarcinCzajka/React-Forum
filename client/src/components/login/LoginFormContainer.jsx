import React from 'react';
import { Menu } from 'semantic-ui-react';
import UserContext from '../../contexts/UserContext';
import LoginForm from './form/LoginForm';
import login from './login';
import handleLogout from './logout';
import { LocaleConsumer } from '../../contexts/LocaleContext';

class LoginFormContainer extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            'open': false,
            'errorMessage': ''
        }
    }

    static contextType = UserContext;
    
    componentDidMount() {
        //Append function that will display LoginForm to context
        this.context.addShowLoginToContext(this.open);
    }

    handleLogin = async (data) => {
        const { email, password, stayLogged } = data;

        await login(email, password)
            .then(result => {
                this.context.setContextData({...result, loggedIn: true});
                
                if(stayLogged) localStorage.setItem('token', document.cookie);

                this.setState({
                    'open': false,
                    'email': '',
                    'password': '',
                    'errorMessage': ''
                })
            })
            .catch(err => {
                this.setState({ errorMessage: err })
            })
    }

    showSignup = () => {
        //Hide LoginForm and display SignUpForm
        this.hideLoginForm();
        this.context.showSignup();
    }

    showLoginForm = () => {
        this.setState({open: true});
    }

    hideLoginForm = () => {
        this.setState({open: false});
    }

    render() {
        return (
            <LocaleConsumer>
                {locale => (
                    !this.context.loggedIn ? (
                        <>
                            <Menu.Item onClick={this.showLoginForm} name={locale.login.loginButton} />
                            <LoginForm 
                                open={this.state.open}
                                onClose={this.hideLoginForm}
                                handleLogin={this.handleLogin}
                                errorMessage={this.state.errorMessage}
                                showSignup={this.showSignup}
                            />
                        </>
                    ) : (
                        <Menu.Item onClick={() => handleLogout(this.context)} name={locale.login.logoutButton} />
                    )
                )}
            </LocaleConsumer>
        )
    }

}

export default LoginFormContainer;