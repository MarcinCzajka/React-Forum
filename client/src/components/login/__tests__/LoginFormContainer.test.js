import React from 'react';
import { create } from "react-test-renderer";
import { shallow } from 'enzyme';
import LoginFormContainer from '../LoginFormContainer';
import { UserProvider } from '../../../contexts/UserContext';

const context = (loggedIn) => {
    return {
        loggedIn: loggedIn,
        addShowLoginToContext: function(){},
        showSignup: function(){}
    }
}

describe('LoginFormContainer component', () => {
    test('renders without crashing', () => {
        shallow(<UserProvider value={context('')}><LoginFormContainer /></UserProvider>);
    });

    test('displays correct button when logged in', () => {
        const component = create(<UserProvider value={context(true)}><LoginFormContainer /></UserProvider>);
        expect(component).toMatchSnapshot();
    });

    test('displays correct button when logged out', () => {
        const component = create(<UserProvider value={context(false)}><LoginFormContainer /></UserProvider>);
        expect(component).toMatchSnapshot();
    });

});