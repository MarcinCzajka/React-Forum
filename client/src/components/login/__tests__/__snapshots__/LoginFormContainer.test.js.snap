// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginFormContainer component displays correct button when logged in 1`] = `
<a
  className="item"
  onClick={[Function]}
>
  Log Out
</a>
`;

exports[`LoginFormContainer component displays correct button when logged out 1`] = `
<a
  className="item"
  onClick={[Function]}
>
  Log In
</a>
`;
